src/shared/copy.c

```diff
@@ -275,9 +275,17 @@ int copy_bytes_full(
                         /* If we're in a hole (current offset is not a data offset), create a hole of the
                          * same size in the target file. */
                         if (e > c) {
-                                r = create_hole(fdt, e - c);
+                                /* Make sure the hole does not exceed the maximum allowed size */
+                                n = MIN(max_bytes, (size_t) e - c);
+                                r = create_hole(fdt, n);
                                 if (r < 0)
                                         return r;
+
+                                if (max_bytes <= (uint64_t) n)
+                                        break;
+
+                                /* Note how many bytes were transferred as required */
+                                max_bytes -= max_bytes == UINT64_MAX ? 0 : n;
```

Comment:
How about this instead:

```
if (e > c) {
        /* Make sure our new hole doesn't go over the maximum size we're allowed to
         * copy. */
        n = MIN(max_bytes, (uint64_t) e - c);

        r = create_hole(fdt, n);
        if (r < 0)
                return r;

        /* Make sure holes are taken into account in the maximum size we're supposed
         * to copy. */
        max_bytes -= n;
        if (max_bytes <= 0)
                break;

        /* Update the size we're supposed to copy in this iteration if needed. */
        if (m > max_bytes)
                 m = max_bytes;
}
```

Reply:
The first part certainly helps, especially for 32-bit, but the second is (I think) incorrect. max_bytes should stay as UINT64_MAX (i.e.: not defined) if the called does not specify a value. That is why the ternary operator is there. I only wanted to change max_bytes at the last moment.

I'm more than open to suggestions as long as the tests pass. I think the mkosi test is possibly unrelated.

Reply:
Right, good point, this should solve that issue

```
if (e > c) {
        /* Make sure our new hole doesn't go over the maximum size we're allowed to
         * copy. */
        n = MIN(max_bytes, (uint64_t) e - c);

        r = create_hole(fdt, n);
        if (r < 0)
                return r;

        /* Make sure holes are taken into account in the maximum size we're supposed
         * to copy. */
        if (max_bytes != UINT64_MAX) {
                max_bytes -= n;
                if (max_bytes <= 0)
                        break;
        )

        /* Update the size we're supposed to copy in this iteration if needed. */
        if (m > max_bytes)
                 m = max_bytes;
}
```

Reply:
Done